### PR TITLE
Make MessageDB versions 0-based rather than -1 based

### PR DESCRIPTION
**Pull request by nordfjord**
Follow up from #339

### CODE REVIEW

File: src/Equinox.MessageDb/MessageDb.fs [hunk 1]
```diff
@@ -257,8 +257,10 @@ module private Read =
 module private Token =
     let create streamVersion : StreamToken =
```
Comment (bartelink) [inline src/Equinox.MessageDb/MessageDb.fs:258]: `// NOTE MessageDB's streamVersion is -1 based, similar to ESDB and SSS`

File: src/Equinox.MessageDb/MessageDb.fs [hunk 2]
```diff
@@ -257,8 +257,10 @@ module private Read =
...
         { value = box streamVersion
-          version = streamVersion
+          version = streamVersion + 1L
```
Comment (bartelink) [inline src/Equinox.MessageDb/MessageDb.fs:260]: // The `Version` exposed on the `ISyncContext` is 0-based

File: src/Equinox.MessageDb/MessageDb.fs [hunk 3]
```diff
@@ -257,8 +257,10 @@ module private Read =
...
           streamBytes = -1 }
+    let inline version (token: StreamToken) = unbox<int64> token.value
```
Comment (bartelink) [inline src/Equinox.MessageDb/MessageDb.fs:262]: I think `streamVersion` is a better name - `version` in Equinox and the Cosmos and DDB stores is 0-based (naming is more terse as 1. Equinox is only concerned with streams 2. to differentiate it from StreamVersion/StreamRevision or other terms used by stores)
`streamVersion` in SSS/ESDB is -1 based

File: src/Equinox.MessageDb/MessageDb.fs [hunk 4]
```diff
@@ -292,12 +294,14 @@ type MessageDbContext(connection : MessageDbConnection, batchOptions : BatchOpti
         return Token.create version, Array.chooseV tryDecode events }
     member _.Reload(streamName, requireLeader, log, token, tryDecode)
         : Async<StreamToken * 'event array> = async {
-        let streamPosition = token.version + 1L
+        let tokenVersion = Token.version token
+        let streamPosition = tokenVersion + 1L
```
Comment (bartelink) [inline src/Equinox.MessageDb/MessageDb.fs:298]: maybe rename to `startPos` ? maybe add comment `// streamVersion (incusive) from which we want to load new events`

File: src/Equinox.MessageDb/MessageDb.fs [hunk 5]
```diff
@@ -292,12 +294,14 @@ type MessageDbContext(connection : MessageDbConnection, batchOptions : BatchOpti
...
         let! version, events = Read.loadForwardsFrom log connection.ReadRetryPolicy connection.Reader batchOptions.BatchSize batchOptions.MaxBatches streamName streamPosition requireLeader
-        return Token.create (max token.version version), Array.chooseV tryDecode events }
+        return Token.create (max tokenVersion version), Array.chooseV tryDecode events }
 
     member _.TrySync(log, streamName, token, encodedEvents : IEventData<EventBody> array): Async<GatewaySyncResult> = async {
-        match! Write.writeEvents log connection.WriteRetryPolicy connection.Writer streamName token.version encodedEvents with
+        let version = Token.version token
```
Comment (bartelink) [inline src/Equinox.MessageDb/MessageDb.fs:303]: streamVersion?

File: tests/Equinox.EventStore.Integration/StoreIntegration.fs [hunk 6]
```diff
@@ -446,3 +458,19 @@ type Tests(testOutputHelper) =
...
+        let! connection = connectToLocalStore log
+
+        let batchSize = 3
+        let context = createContext connection batchSize
+        let id = $"{Guid.NewGuid():N}"
+        let decider = SimplestThing.resolve log context struct("SimplestThing", id)
+
+        let! before, after = decider.TransactEx(
+            (fun state -> state.Version, [SimplestThing.StuffHappened]),
+            mapResult = (fun s result -> s, result.Version))
```
Comment (bartelink) [inline tests/Equinox.EventStore.Integration/StoreIntegration.fs:473]: s and result need to be flipped (`s` is like `state` in your first `fun`, but updated; `result` is the `state.Version` you yielded from the first `fun`)

File: tests/Equinox.EventStore.Integration/StoreIntegration.fs [hunk 7]
```diff
@@ -446,3 +458,19 @@ type Tests(testOutputHelper) =
...
+        test <@ before = 0L @>
```
Comment (bartelink) [inline tests/Equinox.EventStore.Integration/StoreIntegration.fs:474]: I tend to combine these with && so a fail will give more context

File: tests/Equinox.EventStore.Integration/StoreIntegration.fs [hunk 8]
```diff
@@ -89,12 +89,24 @@ type Category<'event, 'state, 'context> = EventStoreCategory<'event, 'state, 'co
 
 let createContext connection batchSize = Context(connection, batchSize = batchSize)
 
+module SimplestThing =
```
Comment (bartelink) [inline tests/Equinox.EventStore.Integration/StoreIntegration.fs:92]: nice!

File: tests/Equinox.EventStore.Integration/StoreIntegration.fs (hunk 7 above)
Comment (nordfjord) [reply tests/Equinox.EventStore.Integration/StoreIntegration.fs]: I did `<@ [before; after] = [0L; 1L] @>` for terseness

File: tests/Equinox.EventStore.Integration/StoreIntegration.fs (hunk 6 above)
Comment (nordfjord) [reply tests/Equinox.EventStore.Integration/StoreIntegration.fs]: yessir

**Comment by bartelink** [comment]
Looks great, thanks! Shout now if you don't want me to merge it when it's done

**Comment by nordfjord** [comment]
Yeah, I think it's ready from my perspective. I ran the tests against SqlStreamStore's PG as well to be safe and it's all green

**Pull request merged**